packages/mui-base/src/Tabs/TabIndicator/TabIndicator.tsx

```diff
@@ -0,0 +1,150 @@
+'use client';
+import * as React from 'react';
+import PropTypes from 'prop-types';
+import { useTabIndicator } from './useTabIndicator';
+import { useTabIndicatorStyleHooks } from './useTabIndicatorStyleHooks';
+import { TabIndicatorProps } from './TabIndicator.types';
+import { useTabsContext } from '../Root/TabsContext';
+import { resolveClassName } from '../../utils/resolveClassName';
+import { defaultRenderFunctions } from '../../utils/defaultRenderFunctions';
+import { evaluateRenderProp } from '../../utils/evaluateRenderProp';
+import { useRenderPropForkRef } from '../../utils/useRenderPropForkRef';
+
+const TabIndicator = React.forwardRef<HTMLSpanElement, TabIndicatorProps>(
+  function TabIndicator(props, forwardedRef) {
+    const {
+      className: classNameProp,
+      render: renderProp,
+      renderBeforeHydration = false,
+      ...other
+    } = props;
+    const render = renderProp ?? defaultRenderFunctions.span;
+    const [instanceId] = React.useState(() => Math.random().toString(36).slice(2));
+    const [isMounted, setIsMounted] = React.useState(false);
+    const { value: activeTabValue } = useTabsContext();
+
+    React.useEffect(() => {
+      setIsMounted(true);
+    }, []);
+
+    const {
+      direction,
+      getRootProps,
+      orientation,
+      activeTabPosition: selectedTabPosition,
+      tabActivationDirection,
+    } = useTabIndicator();
+
+    const ownerState = {
+      selectedTabPosition,
+      orientation,
+      direction,
+      tabActivationDirection,
+    };
+
+    const className = resolveClassName(classNameProp, ownerState);
+    const styleHooks = useTabIndicatorStyleHooks(ownerState);
+    const mergedRef = useRenderPropForkRef(render, forwardedRef);
+
+    if (activeTabValue == null) {
+      return null;
+    }
+
+    const rootProps = getRootProps({
+      ...styleHooks,
+      ...other,
+      className,
+      ref: mergedRef,
+      'data-instance-id': !(isMounted && renderBeforeHydration) ? instanceId : undefined,
+      suppressHydrationWarning: true,
+    } as React.ComponentPropsWithRef<'span'>);
+
+    // This script is used to set the initial position of the indicator before hydration happens.
+    // This is necessary to render the indicator in the right place right after the SSR-generated content is downloaded and not wait for React to kick in.
+    const prehydrationScript = `
+(function() {
+  let indicator = document.currentScript.previousElementSibling;
+  if (!indicator) {
```

Comment:
I think it may be better to store the source of this in its own file (unused) and use a minified version here?